tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -144,6 +153,11 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
                 len++;
             }
 
+            if (span.SpanLinks is { Count: > 0 })
```

Comment:
As we saw during your testing, there is a race condition between this `if (span.SpanLinks is { Count: > 0 })` and the one in line 218 below. Since serialization happens in a separate thread, another piece of code could add or remove span links between these two lines and we could end with the wrong number of fields in `len`.

Maybe we can do something similar to the `hasMetaStruct` local variable above:
```csharp
bool hasSpanLinks = span.SpanLinks is { Count: > 0 };

if (hasSpanLinks) // change both checks
{
...
```

This way, if there were no span links and someone adds one during serialization, we ignore the new one. If there were spans and someone _removes_ them, I think we'll end up adding an empty array, which should be fine. 

There is still the issue with someone adding or removing span links during the `foreach` below. Modifying the list while it's being iterated will thrown an exception. We may need actual lock to avoid that issue. 🤔 

(btw, there is a similar issue `span.Error` above, something to fix in a separate PR someday...)

Reply:
All of this also applies to the span link attributes.

Reply:
changed!